control-plane-operator/controllers/hostedcontrolplane/mcs/reconcile.go

```diff
@@ -0,0 +1,143 @@
+)
+
+func ReconcileMachineConfigServerConfig(cm *corev1.ConfigMap, p *MCSParams) error {
+	p.OwnerRef.ApplyTo(cm)
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	serializedDNS, err := serialize(p.DNS)
+	if err != nil {
+		return err
+	}
+	serializedInfra, err := serialize(p.Infrastructure)
+	if err != nil {
+		return err
+	}
+	serializedNetwork, err := serialize(p.Network)
+	if err != nil {
+		return err
+	}
+	serializedProxy, err := serialize(p.Proxy)
+	if err != nil {
+		return err
+	}
+	serializedPullSecret, err := serialize(p.PullSecret)
+	if err != nil {
+		return err
+	}
+	serializedMasterConfigPool, err := serializeConfigPool(masterConfigPool())
+	if err != nil {
+		return err
+	}
+	serializedWorkerConfigPool, err := serializeConfigPool(workerConfigPool())
+	if err != nil {
+		return err
+	}
+
+	cm.Data["root-ca.crt"] = string(p.RootCA.Data[pki.CASignerCertMapKey])
+	cm.Data["combined-ca.crt"] = string(p.CombinedCA.Data[pki.CASignerCertMapKey])
+	cm.Data["cluster-dns-02-config.yaml"] = serializedDNS
+	cm.Data["cluster-infrastructure-02-config.yaml"] = serializedInfra
+	cm.Data["cluster-network-02-config.yaml"] = serializedNetwork
+	cm.Data["cluster-proxy-01-config.yaml"] = serializedProxy
+	cm.Data["install-config.yaml"] = p.InstallConfig.String()
+	cm.Data["pull-secret.yaml"] = serializedPullSecret
+	cm.Data["master.machineconfigpool.yaml"] = serializedMasterConfigPool
+	cm.Data["worker.machineconfigpool.yaml"] = serializedWorkerConfigPool
+	return nil
+}
+
+func serialize(obj client.Object) (string, error) {
+	return util.SerializeResource(obj, api.Scheme)
+}
+
+func serializeConfigPool(obj client.Object) (string, error) {
+	return util.SerializeResource(obj, &machineConfigTyper{})
+}
+
+// NOTE: this is here because we don't import the the machineconfig types into
+// our scheme, and apart from this use case, we don't need to.
+type machineConfigTyper struct{}
+
+func (t *machineConfigTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind, bool, error) {
+	return []schema.GroupVersionKind{
+		{
+			Group:   "machineconfiguration.openshift.io",
+			Version: "v1",
+			Kind:    "MachineConfigPool",
+		},
+	}, false, nil
+}
+
+func (t *machineConfigTyper) Recognizes(gvk schema.GroupVersionKind) bool {
+	return true
+}
+
+func masterConfigPool() *mcfgv1.MachineConfigPool {
+	return &mcfgv1.MachineConfigPool{
+		ObjectMeta: metav1.ObjectMeta{
+			// NOTE: this configpool is master in name only but is used to
+			// render config for workers in a hypershift cluster.
+			Name: "master",
+			Labels: map[string]string{
+				"machineconfiguration.openshift.io/mco-built-in": "",
+			},
+		},
+		Spec: mcfgv1.MachineConfigPoolSpec{
+			MachineConfigSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					// NOTE: the master config pool is the only pool that
+					// can be served by the MCS in bootstrap mode. For the
```

Comment:
Can we update the MCS to not end up creating a non-master pool named master?

Reply:
We should. But this is what currently works. Maybe we can modify the MCS to allow specifying the name of the machineconfigpool that can be served in bootstrap mode.